### ISSUE TITLE
[Question] If a domain exists in >=2 adlists, each of which have no unique domains, will the domain no longer be blocked (if told to remove adlists without unique ones)?

### ISSUE THREAD

**Issue by Ghalid** [yubiuser/pihole_adlist_tool#5]
The question in the title may be phrased a little oddly, but I'll try to explain it better here. Take the following example scenario:

I have 5 adlists. For simplicity's sake, I'll just reference them by their id 1-5.

Let's use the following data for these 5 adlists (the specific numbers don't really matter, just simply whether a domain/unique domain is present or not):

id | domains_covered | unique_domains_covered
-- | ----------------- | -------------------------
1 | 24 | 3
2 | 43 | 6
3 | 1 |  
4 | 1 |  
5 |     |   

Now, let's say I choose to "Enable only adlists with covered unique domains" on the final question. From my understanding, only adlists 1 and 2 will remain enabled, while the rest will be disabled.

Notice how lists 3 and 4 have one covered domain. Let's say this domain is `example.com`. `example.com` exists in both list 3 and 4, but not any of the other lists. I assume that this would mean that neither list 3 nor list 4 are considered to have a unique covered domain.

Would it be correct to say that, if you were to let the script enable only adlists containing unique domains, that protection against the `example.com` domain would be lost? Or is there something in the script that accounts for this that I'm missing? I feel that for people who use a large amount of adlists, this could be a fairly common "special case."

More importantly, if `example.com` does end up being lost, is there an easy way to determine a list of domains like this one that would be lost by skimming the adlists, in a real scenario where you don't know the all the specific domains in each adlist?

**Comment by yubiuser** [comment]
Thanks for your suggestions.

This is something I haven't really thought about. You'r right, `example.com` would be lost in that case. 
I haven't found a good solution to the problem (but if you got one, I would be happy to know about).

I see some problems: How should the script decide which adlist (nr. 3, nr. 4) it should deactivate? It would require a fairly complex mechanism to detect which domains would be lost by deactivating and which is the **best** (minimum?) number of adlists to keep. In a more complex real-world scenario there would be much more adlists containing different non-identical numbers of duplicated domains. I this is a classical optimization problem. Of course I could try every possible combination but this might be not very efficient. 

Another idea would be to disable one adlist (without unique domains) (say nr.3 ) and run the script again. This time nr. 4 would contain an unique domain. But also here: if you have 20 adlists with no unique domain, in worst case you would have to run the script 19 times to (if all would contain the exactly same duplicate domains) to reach the "optimum" (minimum number of adlist which cover all visited domains).

To summarize: You're right, we loose some domains choosing "enable only adlists with unique domains". But it is hard to circumvent that. I'll give it some more thoughts. Maybe adding a more explicite info would also be necessary.

**Comment by yubiuser** [comment]
Ok, I've found a solution which will be an iterative reduction of `gravity_strip`. It will cost some performance, but in general should still be fast. 

But I'll need some time to write the code.

**Comment by Ghalid** [comment]
> Ok, I've found a solution which will be an iterative reduction of `gravity_strip`. It will cost some performance, but in general should still be fast.
> 
> But I'll need some time to write the code.

Awesome, thanks! Take as much time as you need.

I think you're thinking of a similar solution, but what I was thinking of:
1. Make a new table, which is basically `gravity_strip` minus `unique_domains` (we'll call it `nonunique_domains` here) that contains only domains that aren't unique to one adlist
2. Search the adlists that already contain unique domains for the `nonunique_domains` - if they are contained in at least one, they can be removed from the table since they don't need to be considered (the current system would still maintain these domains without any changes)
3. Search the adlists that do NOT contain unique domains but have at least one covered domain, and determine how many of the remaining `nonunique_domains` each one has. Choose the one with the greatest amount, or if tied, choose based off of something else (the lower id?).
4. Add the adlist found in step 3 to the list to enable, remove all the `nonunique_domains` covered by this adlist from the table, and repeat step 3 until no domains are left. Of course, for optimization, the list added in this step doesn't need to be searched again.

I don't know whether this will always guarantee the most efficient (least adlists for greatest protection) solution though. If some combination of lists with a lower number of `nonunique_domains` covers all the domains in the list that the script determined to have a higher number of `nonunique_domains` earlier - then you might have "unnecessary" lists enabled. Example: If you have 3 adlists, and list 2 + 3 combined would contain all the domains in list 1 plus more, but list 1 has the greatest amount of `nonunique_domains`, all 3 lists would probably be enabled but you don't actually need list 1.

Another potential solution could be to trim all the adlists that would be searched in step 3 (no unique domains, but with at least one covered domain) to only contain `nonunique_domains` one time, then start trying combinations of those trimmed adlists (combinations containing smallest amount of adlists to largest) until you find one that contains all of the `nonunique_domains`. I don't know what the performance impact of this would be though - I assume that the trimming the adlists would make the process take a lot less time (except for the initial trimming, of course).

Sorry for the brain dump, but hopefully something here was of use to you, haha

**Comment by yubiuser** [comment]
We had exactly the same idea :)

I would phrase it like this (more for me to remember and know how to code it :-)

0) Disable all adlists.

(Repeat until `gravity_strip` is empty) (not really this table, will be an copy)
1) Determine `unique domains` and the adlists they are on.
If there are unique domains:
 1.1) Enable adlist (with unique domains)
1.2) Remove all domains (unique and non-unique) from `gravity_strip` that are contained on adlists which contain unique domains.
2) Count how many domains remain for each adlist.
3) Remove all domains that are on the adlist with the highest number of domains. 
If two list contain the same amount, choose random.
4) Enable this particular adlist.
5) Start with 1) over

Check that `blocked_domains` are contained completely in the adlists that are enabled at the end.

____


>I don't know whether this will always guarantee the most efficient (least adlists for greatest protection) solution though. If some combination of lists with a lower number of nonunique_domains covers all the domains in the list that the script determined to have a higher number of nonunique_domains earlier - then you might have "unnecessary" lists enabled. Example: If you have 3 adlists, and list 2 + 3 combined would contain all the domains in list 1 plus more, but list 1 has the greatest amount of nonunique_domains, all 3 lists would probably be enabled but you don't actually need list 1.

I can't construct such a case without list 2 and 3 containing unique domains even before removing (enabling) list 1 - maybe you could make a concrete example?


1 | 2 | 3
-- | -- | --
a | a | c
b | b | d
c |   |  
d |   | 
  |   |  |g
  | h |  

(Only if list 2+3 contain the **exactly identical** additional  domains. Then list 1 and one of list 2 or 3 would be enabled)

1 | 2 | 3
-- | -- | --
a | a | c
b | b | d
c |   |  
d |   | 
  |   |  |h
  | h |  

> guarantee the most efficient (least adlists for greatest protection) 

Depends on what you consider "efficient". If you go for "greatest protection" you could just enable every adlist you find on the net - it will probably protect you form more then you want. I would argue most efficient is a small number of high-quality adlists that cover all your actual use cases.

> Another potential solution could be to trim all the adlists that would be searched in step 3 (no unique domains, but with at least one covered domain)

As the idea was to use `gravity_strip`, this is not necessary as it contains already only the domains (and adlists) with covered (visited) domains. If I understand you correctly, this is identical with your 
>except for the initial trimming, of course
___
> Another potential solution 

I'm not sure if trying all combinations (pairs, trippletts, quartetts, ...) to find a combination that covers all `nonunique_domains` (bottom-up) would be faster than the top-down approach you described earlier. I consider counting + comparing way faster than permutating through all combinations (expect users to have 20 instead of only 3 adlists)

**Comment by Ghalid** [comment]
> I can't construct such a case without list 2 and 3 containing unique domains even before removing (enabling) list 1 - maybe you could make a concrete example?

You're right, I had an error in logic. I wasn't considering the unique domains (which is a pretty important factor, lol).

In that case, completely ignore my second solution 😛 

> Depends on what you consider "efficient". If you go for "greatest protection" you could just enable every adlist you find on the net - it will probably protect you form more then you want. I would argue most efficient is a small number of high-quality adlists that cover all your actual use cases.

Probably should've clarified that by "greatest protection" I meant protection for your actual use cases. So yeah, we basically agree completely

> As the idea was to use `gravity_strip`, this is not necessary as it contains already only the domains (and adlists) with covered (visited) domains. If I understand you correctly, this is identical with your

My point was to trim adlists with extremely large amounts of domains, many of which you have and never will see, to only the domains in the `gravity_strip` table, so that they would be potentially faster to process if you had to try many combinations. For example, for an adlist with 200,000 domains on it, you'd trim it down to say, 100 domains (if that many domains match the ones in `gravity_strip`), and use that trimmed adlist to try combinations instead. However, now that I think about it, this process would probably be useless, and going with the most obvious solution would be better (the first top-down approach rather than the "try every combination and hope one sticks")

**Comment by yubiuser** [comment]
> that trimmed adlist to try combinations instead.

Yeah, that's what I do by using only domains from `gravity_strip`. I would never try all combinations from all domains on the adlists, but rather those on `gravity_strip`. Example: Adlist 1 contains 1 million domains, but only 100 have been visited. List 2 contains 2 million domains but only 50 have been visited. Gravity_strip contains only these 100+50, which I would use for comparison.

**Comment by yubiuser** [comment]
Short follow up. 

I started from the end and wrote the part of
>Check that blocked_domains are contained completely in the adlists that are enabled at the end.

For my configuration, it looks like this when first enabling all adlists to analyze and then selecting only adlists with unique domains

```
You have 37 adlists configured (37 enabled)
  [i]  Your gravity.db contains 1288657 unique domains
  [i]  Since 27.07.2020 23:11:20 799 different domains from your adlists have been blocked 62557 times in total
       (blocked directly by gravity or during deep CNAME inspection)
  [i]  Using you current adlist configuration 872 domains would have been blocked 73606 times

[...]

[i]  In total your adlists contain 156 visited (covered) unique domains - meaning those domains are contained only in a single adlist.

[...]

[✓]  Adlists with covered unique domains enabled


[✓]  Of the 872 domains that would have been blocked by the adlist configuration you chose to analyze,
       710 domains are covered by your current adlist configuration.
```

So I would looe 162 domains due to the issue you raised here.

**Comment by yubiuser** [comment]
Playing with it, I came to the conclusion that

>(Repeat until gravity_strip is empty) (not really this table, will be an copy)
>
>1.Determine unique domains and the adlists they are on.
>If there are unique domains:
>1.1) Enable adlist (with unique domains)
>1.2) Remove all domains (unique and non-unique) from gravity_strip that are contained on adlists which contain unique >domains.
>2.Count how many domains remain for each adlist.
>3.Remove all domains that are on the adlist with the highest number of domains.
>If two list contain the same amount, choose random.
>4.Enable this particular adlist.
>5. Start with 1) over
>Check that blocked_domains are contained completely in the adlists that are enabled at the end.

Can be modified to 

(Repeat until `gravity_strip` is empty) (not really this table, will be an copy)
1) Determine `unique domains` and the adlists they are on.
If there are unique domains:
 1.1) Enable adlist (with unique domains)
1.2) Remove all domains (unique and non-unique) from `gravity_strip` that are contained on adlists which contain unique domains.
2) Count how many domains remain for each adlist.
3) Remove all domains that are on the adlist with the highest number of domains. 
If two list contain the same amount, choose the one with the lower adlist_id.
4) Enable this particular adlist.
5) Start with 2) over

There is no need to check for unique domains again after they have been removed initially. If we know that all remaining adlists contain only domains that are found on other adlists there are no unique domains left, even after removing adlists step-by-step.
